### ISSUE TITLE
Inconsistent behavior while updating variable value

### ISSUE THREAD

**Issue by gthyagi** [underworldcode/underworld3-archive#146]
Hi, 

I am noticing inconsistent behavior in updating variable value. 

If I run the attached script with following option
```python3
meshes = [unstructured_quad_box_irregular]
```
outputs `Initial value:  0.9857812808263339`
```
.meshes/uw_simplexbox_minC(-0.5, -0.5)_maxC(0.5, 0.5)_csize0.03125_regFalse.msh
Initial value:  0.9857812808263339
Final value:  0.9767381595400257
-0.009043121286308131 -0.009025777832627232
```

If I run the attached script with following option
```python3
meshes = [structured_quad_box, unstructured_quad_box_irregular, unstructured_quad_box_regular,]
```
outputs `Initial value:  0.9171282849299894`
```
.meshes/uw_structuredQuadBox_minC(-0.5, -0.5)_maxC(0.5, 0.5).msh
Initial value:  0.9735824134442078
Final value:  0.9171282849299894
-0.05645412851421838 -0.060239284674289795
.meshes/uw_simplexbox_minC(-0.5, -0.5)_maxC(0.5, 0.5)_csize0.03125_regFalse.msh
Initial value:  0.9171282849299894
Final value:  0.9776598451615978
0.06053156023160833 -0.009025777832627232
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[7], line 62
     59 print(T_solve[0]-T_0[0], np.float64(dTdt.evalf(subs={x: x_0[0], y: x_0[1]})*time))
     61 if not np.allclose(T_solve[0]-T_0[0], np.float64(dTdt.evalf(subs={x: x_0[0], y: x_0[1]})*time), atol=1e-2):
---> 62     raise RuntimeError('Analytical and numerical solution not close')
     64 # del mesh
     65 # del adv_diff, v, T
     66 
     67 # return

RuntimeError: Analytical and numerical solution not close
```
code is failing to update the value here:
```python3
# storing initial temperature value in a array
T_0 = np.zeros((1, 1))
T_0 = uw.function.evaluate(T.sym[0], np.array([x_0]))
```

T values are resetting to zero in each iteration. Despite that it produces `Initial value:  0.9171282849299894` instead of `Initial value:  0.9857812808263339`.

[test1003_Ex_AdvDiff_rotatingCone.py.txt](https://github.com/underworldcode/underworld3/files/13874902/test1003_Ex_AdvDiff_rotatingCone.py.txt)

**Comment by lmoresi** [comment]
Can we check where this problem originates with a more targeted example script ?  Is it assigning values that goes wrong, evaluating the initial values, or in retrieving the values for analysis.

**Comment by gthyagi** [comment]
I believe the root of this problem lies in reusing the same variable symbol to represent the mesh variable. I encountered a similar issue recently while plotting two different mesh variables on different meshes but using the same symbol in PyVista. I will investigate further into this matter and provide an update.

**Comment by gthyagi** [comment]
From my tests, it was verified that the problem stemmed from the repeated use of the mesh variable symbol.

Solution: Create different mesh variable symbol each time. This can be achieved by following these steps.

```python3
@pytest.mark.parametrize("mesh, v_symb, T_symb",
                         [(structured_quad_box, "U_sqb", "T_sqb"),
                          (unstructured_quad_box_irregular, "U_iqbir", "T_iqbir"),
                          (unstructured_quad_box_regular, "U_iqbre", "T_iqbre")],
                        )

def test_adv_diff_rotating_cone(mesh, v_symb, T_symb):
    # Create an mesh vars
    v = uw.discretisation.MeshVariable(v_symb, mesh, mesh.dim, degree=2)
    T = uw.discretisation.MeshVariable(T_symb, mesh, 1, degree=1)

```

**Comment by lmoresi** [comment]
We split the variable name (used by petsc) from the variable symbol (used by sympy).

It might just be fine to make the former genuinely unique by taking it out of the user control.


Sent from my iPhone

On 26 Feb 2024, at 00:08, Thyagarajulu Gollapalli ***@***.***> wrote:

﻿

Closed #146<https://github.com/underworldcode/underworld3/issues/146> as completed.

**Comment by gthyagi** [comment]
I agree, variable symbols are mainly used for displaying equations in a tidy format. It will be make users life easy if we implement unique variable symbol with each mesh call.
